### PR TITLE
Feature/progress bar

### LINKED ISSUE #26: Progress Bar

### PR DESCRIPTION
**Pull request by thispastwinter**
## Description

This component is intended to be used on processes that take longer than what a user would expect for a typical loading state. Examples include uploading, downloading, or even a page load that takes longer than what is typically expected. For traditional loading state use cases, it is recommended you use the `<Loader />` component.

### 🔗 Reference

> _Relevant links (e.g. issue, design, etc.)_

- [LINK_NAME](LINK_URL)

### 📋 Acceptance Criteria

> _Checked off by PR **Assignees**_

- [ ] CRITERION_HERE

### Determinate

https://user-images.githubusercontent.com/46500734/182413226-16551f4b-2eb3-4750-98bf-e56e13f5c7c6.mov

### Indeterminate

https://user-images.githubusercontent.com/46500734/182413276-0f6693e8-f299-4909-831e-a6197329cd00.mov

### 👩‍🔬 Test Instructions

> _Provided by PR **Assignees**_

1. TEST_STEP_HERE

## 🔎 Reviewer Checklist

> _Checked off by PR **Reviewers**_

- [ ] Merge destination is correct
- [ ] Code is correct as understood and conforms to quality standards
- [ ] Tests have been added where appropriate (unit, visual, end-to-end)
- [ ] Acceptance Criteria have been met

---

> ### Roles & Responsibilities
>
> #### 👨‍💻 Assignee
>
> - Initiator of this PR (be sure to set in GitHub UI)
> - Addresses feedback and change requests
> - Merges PR once approved (usually deletes branch unless `develop` or `release`)
>
> #### 👩‍💻 Reviewer
>
> - Invited to review PR by **Assignee** (via GitHub UI)
> - Is expected to complete a review and address followup

### CODE REVIEW

File: src/components/ProgressBar/ProgressBar.test.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    // Assert
+    expect(receivedIndicator).not.toHaveClass("progressBarIndeterminate")
+    expect(receivedContainer).toHaveAttribute(ariaValueNow, String(value))
+    expect(receivedContainer).toHaveAttribute(ariaValueMin, min)
+    expect(receivedContainer).toHaveAttribute(ariaValueMax, max)
+  })
+  it("has dynamic transform based on determinate value", () => {
+    // Arrange
+    const initialValue = 0
```
Comment (albatrocity) [inline src/components/ProgressBar/ProgressBar.test.tsx:49]: Without getting too crazy with test cases, it might be good to include a test for a non-zero value as well, just to prove this is actually doing math.

File: src/components/ProgressBar/ProgressBar.module.css [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    transform: translateX(-100%);
+    width: 100%;
+  }
+  100% {
+    transform: translateX(200%);
+    width: 50%;
+  }
+}
+
+.progressBarIndeterminate {
```
Comment (roblafeve) [inline src/components/ProgressBar/ProgressBar.module.css]: With css modules, we can utilize simpler names since "ProgressBar" will be prepended. Consider using contextual names like "indeterminate", "indicator", "track", etc.

File: src/components/ProgressBar/ProgressBar.module.css [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  transition: transform 0.2s linear;
+  background: green;
+}
+
+.progressBarTrack {
+  border-radius: 2px;
+  overflow: hidden;
+  height: 12px;
+  position: relative;
+  z-index: -1;
```
Comment (albatrocity) [inline src/components/ProgressBar/ProgressBar.module.css]: Any potential issues with a `-1` zIndex here, given the parent's position isn't specified? Why not `0` here and a `1` for the indicator? Just wondering if the track might disappear if parent elements have a defined positive `zIndex` or something.

File: src/components/ProgressBar/ProgressBar.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  title: "Components/ProgressBar",
+}
+
+const IndeterminateTemplate: ComponentStory<typeof ProgressBar> = (args) => (
+  <ProgressBar {...args} />
+)
+
+export const Indeterminate = IndeterminateTemplate.bind({})
+
+const DeterminateTemplate: ComponentStory<FC<{ interval: number }>> = (
```
Comment (albatrocity) [inline src/components/ProgressBar/ProgressBar.stories.tsx:22]: This is a great visual for a "real-world" scenario. I'd recommend also adding a story of a standalone determinate instance with Storybook controls, just so a dev can check that `50` is actually halfway.

**Review by albatrocity — CHANGES_REQUESTED** [review]

File: src/components/ProgressBar/ProgressBar.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import classNames from "classnames"
+import styles from "./ProgressBar.module.css"
+
+const MIN = 0
+const MAX = 100
+
+const getTransformStyle = (value?: number) => {
+  if (value || value === 0) {
+    const transform = value - 100
```
Comment (roblafeve) [inline src/components/ProgressBar/ProgressBar.tsx]: I see some values (here and elsewhere) that are the same as the constants above. Would it make sense to utilize these throughout?
```suggestion
  if (value || value === MIN) {
    const transform = value - MAX
```

File: src/components/ProgressBar/ProgressBar.test.tsx (hunk 1 above)
Comment (thispastwinter) [reply src/components/ProgressBar/ProgressBar.test.tsx:49]: Hahaha. You caught me being lazy!

File: src/components/ProgressBar/ProgressBar.module.css (hunk 3 above)
Comment (thispastwinter) [reply src/components/ProgressBar/ProgressBar.module.css]: Ah, yeah, that was leftover from a previous styling attempt when using absolute positioning. Good eye!

File: src/components/ProgressBar/README.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
+# `<ProgressBar />`
+
+This component is intended to be used on processes that take longer than what a user would expect for a typical loading state. Examples include uploading, downloading, or even a page load that takes longer that what is typically expected. For traditional loading state use cases, it is recommended you use the [`<Loader />`](../Loader/README.md) component.
```
Comment (roblafeve) [inline src/components/ProgressBar/README.md]:
```suggestion
This component is intended to be used on processes that take longer than what a user would expect for a typical loading state. Examples include uploading, downloading, or even a page load that takes longer than what is typically expected. For traditional loading state use cases, it is recommended you use the [`<Loader />`](../Loader/README.md) component.
```

**Review by roblafeve — COMMENTED** [review]
At first pass, this looks good!

I'll let others speak into it, but here are a few things I noticed.

**Review by roblafeve — APPROVED** [review]

**Comment by thispastwinter** [comment]
Alright @roblafeve approved, I'd still like to get @albatrocity to review changes from his feedback and I'm happy to give an opportunity to @kage1414 to peek and weigh in.

**Review by albatrocity — APPROVED** [review]

File: src/components/ProgressBar/ProgressBar.tsx [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+const MAX = 100
+
+const getTransformStyle = (value?: number) => {
+  if (value || value === MIN) {
+    const transform = value - MAX
+    return `translateX(${transform}%)`
+  } else {
+    return
+  }
+}
```
Comment (luiz-chagas) [inline src/components/ProgressBar/ProgressBar.tsx:14]: Food for thought:
This function just returns `undefined` if `value` isn't provided (aka, does nothing), if you type this more strictly so `value` _has_ to exist, you get rid of this if/else and the implementation becomes a lot* simpler. The conditional check goes up to the caller, who is most likely be handling the logic anyway.

* it might not feel like the function becomes a lot simpler but it is actually impossible to test this function in isolation with a single test because of the _if/else_ branch.

File: src/components/ProgressBar/ProgressBar.tsx [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  if (value || value === MIN) {
+    return {
+      "aria-valuenow": value,
+      "aria-valuemin": MIN,
+      "aria-valuemax": MAX,
+    }
+  } else {
+    return {}
+  }
+}
```
Comment (luiz-chagas) [inline src/components/ProgressBar/ProgressBar.tsx:32]: same as above

File: src/components/ProgressBar/ProgressBar.tsx [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+export interface ProgressBarProps {
+  /** A number between 0-100 */
+  value?: number
+  ariaLabel?: string
+}
+
+export function ProgressBar({ value, ariaLabel }: ProgressBarProps) {
+  const indicatorClasses = getIndicatorClasses(value)
+  const transform = getTransformStyle(value)
+  const ariaLabels = getAriaProgressLabels(value)
```
Comment (luiz-chagas) [inline src/components/ProgressBar/ProgressBar.tsx:43]: This line could then look like this instead:
```tsx
const ariaLabels = value ? getAriaProgressLabels(value) : {};
```

The benefit is that you see all the possible branching logic here in the component instead of hidden inside internal operators.

This is an idea that sometimes bubbles up from the function level to the component level and can reveal problems where there's no defined behavior for the cases where the values don't exist.

**Review by luiz-chagas — APPROVED** [review]
